server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +462,274 @@ public int hashCode() {
+         *
+         * <p>For each CleanupKey in keysToClean, if the readerCacheKeyId is null or the entity is not open,
+         * the shardId of the entity is added to cleanupKeysFromClosedShards. Otherwise, the CleanupKey is added
+         * to cleanupKeysFromOutdatedReaders.
+         *
+         * @param cleanupKeysFromOutdatedReaders A set to hold CleanupKeys with open shard but invalidated readerCacheKeyId.
+         * @param cleanupKeysFromClosedShards A set to hold CleanupKeys of a closed shard.
+         */
+        private void processCleanupKeys(Set<CleanupKey> cleanupKeysFromOutdatedReaders, Set<Object> cleanupKeysFromClosedShards) {
+            writeLock.lock();
+            try {
+                for (Iterator<CleanupKey> iterator = keysToClean.iterator(); iterator.hasNext();) {
+                    CleanupKey cleanupKey = iterator.next();
+                    iterator.remove();
+                    if (cleanupKey.readerCacheKeyId == null || !cleanupKey.entity.isOpen()) {
+                        // null indicates full cleanup, as does a closed shard
+                        cleanupKeysFromClosedShards.add(((IndexShard) cleanupKey.entity.getCacheIdentity()).shardId());
+                    } else {
+                        cleanupKeysFromOutdatedReaders.add(cleanupKey);
+                    }
+                }
+            } finally {
+                writeLock.unlock();
+            }
+        }
+
+        /**
+         * Determines whether a key should be removed from the cache.
+         *
+         * <p>This method checks if the key's shardId is present in the cleanupKeysFromClosedShards set,
+         * indicating that the shard has been closed and the key should be removed. If the shardId is not present,
+         * it checks if the key's readerCacheKeyId is present in the cleanupKeysFromOutdatedReaders set,
+         * indicating that the reader has been invalidated and the key should be removed.
+         *
+         * @param key The key to check for removal.
+         * @param cleanupKeysFromOutdatedReaders A set of CleanupKeys with open shard but invalidated readerCacheKeyId.
+         * @param cleanupKeysFromClosedShards A set of CleanupKeys of a closed shard.
+         * @return true if the key should be removed, false otherwise.
+         */
+        private boolean shouldRemoveKey(Key key, Set<CleanupKey> cleanupKeysFromOutdatedReaders, Set<Object> cleanupKeysFromClosedShards) {
+            readLock.lock();
+            try {
+                if (cleanupKeysFromClosedShards.contains(key.shardId)) {
+                    return true;
+                } else {
+                    Optional<CacheEntity> cacheEntity = cacheEntityLookup.apply(key.shardId);
+                    if (cacheEntity.isPresent()) {
+                        CleanupKey cleanupKey = new CleanupKey(cacheEntity.get(), key.readerCacheKeyId);
+                        return cleanupKeysFromOutdatedReaders.contains(cleanupKey);
+                    }
```

Comment:
>Can there be a case where for outdated readers shardId will be null ?

I can't think of a case. 
But you made a very keen observation. I'll revert this change to avoid changing behaviour.